frontend/javascripts/oxalis/model/bucket_data_handling/abstract_cuckoo_table.ts

```diff
@@ -0,0 +1,303 @@
+import * as THREE from "three";
+import UpdatableTexture from "libs/UpdatableTexture";
+import { getRenderer } from "oxalis/controller/renderer";
+import { createUpdatableTexture } from "oxalis/geometries/materials/plane_material_factory_helpers";
+
+const TEXTURE_CHANNEL_COUNT = 4;
+const DEFAULT_LOAD_FACTOR = 0.25;
+const EMPTY_KEY = 2 ** 32 - 1;
+const EMPTY_KEY_VALUE = 2 ** 32 - 1;
+
+export type SeedSubscriberFn = (seeds: number[]) => void;
+
+let cachedNullTexture: UpdatableTexture | undefined;
+
+export abstract class AbstractCuckooTable<K, V, Entry extends [K, V]> {
+  static ELEMENTS_PER_ENTRY = 4;
+  entryCapacity: number;
+  table!: Uint32Array;
+  seeds!: number[];
+  seedSubscribers: Array<SeedSubscriberFn> = [];
+  _texture: UpdatableTexture;
+  textureWidth: number;
+
+  constructor(textureWidth: number) {
+    this.textureWidth = textureWidth;
+    this._texture = createUpdatableTexture(
+      textureWidth,
+      textureWidth,
+      TEXTURE_CHANNEL_COUNT,
+      THREE.UnsignedIntType,
+      getRenderer(),
+      THREE.RGBAIntegerFormat,
+    );
+
+    // The internal format has to be set manually, since ThreeJS does not
+    // derive this value by itself.
+    // See https://webgl2fundamentals.org/webgl/lessons/webgl-data-textures.html
+    // for a reference of the internal formats.
+    this._texture.internalFormat = "RGBA32UI";
+
+    this.entryCapacity = Math.floor(
+      (textureWidth ** 2 * TEXTURE_CHANNEL_COUNT) / AbstractCuckooTable.ELEMENTS_PER_ENTRY,
+    );
+
+    this.initializeTableArray();
+    // Initialize the texture once to avoid undefined behavior
+    this.flushTableToTexture();
+  }
+
+  static computeTextureWidthFromCapacity(requestedCapacity: number): number {
+    const capacity = requestedCapacity / DEFAULT_LOAD_FACTOR;
+    const textureWidth = Math.ceil(
+      Math.sqrt((capacity * TEXTURE_CHANNEL_COUNT) / AbstractCuckooTable.ELEMENTS_PER_ENTRY),
+    );
+    return textureWidth;
+  }
+
+  static getNullTexture(): UpdatableTexture {
+    if (cachedNullTexture) {
+      return cachedNullTexture;
+    }
+    cachedNullTexture = createUpdatableTexture(
+      0,
+      0,
+      TEXTURE_CHANNEL_COUNT,
+      THREE.UnsignedIntType,
+      getRenderer(),
+      THREE.RGBAIntegerFormat,
+    );
+    cachedNullTexture.internalFormat = "RGBA32UI";
+
+    return cachedNullTexture;
+  }
+
+  private initializeTableArray() {
+    this.table = new Uint32Array(AbstractCuckooTable.ELEMENTS_PER_ENTRY * this.entryCapacity).fill(
+      EMPTY_KEY_VALUE,
```

Comment:
I'm stumbling about this line, since a subclass could decide not to use `EMPTY_KEY_VALUE` and instead return something else in `getEmptyValue()`. If you use that instead, there should be a check whether `ELEMENTS_PER_ENTRY > 1` and if so that all entries of the empty key are the same, before using `getEmptyValue()[0]` to fill the array.

Reply:
You are right! I guess, my generalization of these classes got a bit too flexibel (which wouldn't have worked if the subclasses were to push the limits). I cleaned this up (mainly by making the restrictions clearer in the comments + interface). See 7e4db9587.